// devextreme/exporter/exceljs/export_pivot_grid.js
/**
 * DevExtreme (exporter/exceljs/export_pivot_grid.js)
 * Version: 20.1.7
 * Build date: Tue Aug 25 2020
 *
 * Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.exportPivotGrid = void 0;
var _type = require("../../core/utils/type");
var _export = require("./export");
var _pivot_grid = require("../../ui/pivot_grid");
var _pivot_grid2 = _interopRequireDefault(_pivot_grid);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        "default": obj
    }
}
var privateOptions = {
    _getWorksheetFrozenState: function(dataProvider, cellRange) {
        return {
            state: "frozen",
            xSplit: cellRange.from.column + dataProvider.getFrozenArea().x - 1,
            ySplit: cellRange.from.row + dataProvider.getFrozenArea().y - 1
        }
    },
    _setAutoFilter: function() {},
    _setFont: function() {},
    _trySetOutlineLevel: function() {},
    _getCustomizeCellOptions: function(excelCell, pivotCell) {
        return {
            excelCell: excelCell,
            pivotCell: pivotCell
        }
    },
    _needMergeRange: function() {
        return true
    }
};

function exportPivotGrid(options) {
    return _export.Export.export(_getFullOptions(options), privateOptions)
}

function _getFullOptions(options) {
    if (!((0, _type.isDefined)(options) && (0, _type.isObject)(options))) {
        throw Error('The "exportPivotGrid" method requires a configuration object.')
    }
    if (!((0, _type.isDefined)(options.component) && (0, _type.isObject)(options.component) && options.component instanceof _pivot_grid2.default)) {
        throw Error('The "component" field must contain a PivotGrid instance.')
    }
    return _export.Export.getFullOptions(options)
}
exports.exportPivotGrid = exportPivotGrid;
