// devextreme-react/index.js
/*!
 * devextreme-react
 * Version: 20.1.7
 * Build date: Tue Aug 25 2020
 *
 * Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var template_1 = require("./core/template");
exports.Template = template_1.Template;
var accordion_1 = require("./accordion");
exports.Accordion = accordion_1.Accordion;
var action_sheet_1 = require("./action-sheet");
exports.ActionSheet = action_sheet_1.ActionSheet;
var autocomplete_1 = require("./autocomplete");
exports.Autocomplete = autocomplete_1.Autocomplete;
var bar_gauge_1 = require("./bar-gauge");
exports.BarGauge = bar_gauge_1.BarGauge;
var box_1 = require("./box");
exports.Box = box_1.Box;
var bullet_1 = require("./bullet");
exports.Bullet = bullet_1.Bullet;
var button_1 = require("./button");
exports.Button = button_1.Button;
var button_group_1 = require("./button-group");
exports.ButtonGroup = button_group_1.ButtonGroup;
var calendar_1 = require("./calendar");
exports.Calendar = calendar_1.Calendar;
var chart_1 = require("./chart");
exports.Chart = chart_1.Chart;
var check_box_1 = require("./check-box");
exports.CheckBox = check_box_1.CheckBox;
var circular_gauge_1 = require("./circular-gauge");
exports.CircularGauge = circular_gauge_1.CircularGauge;
var color_box_1 = require("./color-box");
exports.ColorBox = color_box_1.ColorBox;
var context_menu_1 = require("./context-menu");
exports.ContextMenu = context_menu_1.ContextMenu;
var data_grid_1 = require("./data-grid");
exports.DataGrid = data_grid_1.DataGrid;
var date_box_1 = require("./date-box");
exports.DateBox = date_box_1.DateBox;
var defer_rendering_1 = require("./defer-rendering");
exports.DeferRendering = defer_rendering_1.DeferRendering;
var diagram_1 = require("./diagram");
exports.Diagram = diagram_1.Diagram;
var draggable_1 = require("./draggable");
exports.Draggable = draggable_1.Draggable;
var drawer_1 = require("./drawer");
exports.Drawer = drawer_1.Drawer;
var drop_down_box_1 = require("./drop-down-box");
exports.DropDownBox = drop_down_box_1.DropDownBox;
var drop_down_button_1 = require("./drop-down-button");
exports.DropDownButton = drop_down_button_1.DropDownButton;
var file_manager_1 = require("./file-manager");
exports.FileManager = file_manager_1.FileManager;
var file_uploader_1 = require("./file-uploader");
exports.FileUploader = file_uploader_1.FileUploader;
var filter_builder_1 = require("./filter-builder");
exports.FilterBuilder = filter_builder_1.FilterBuilder;
var form_1 = require("./form");
exports.Form = form_1.Form;
var funnel_1 = require("./funnel");
exports.Funnel = funnel_1.Funnel;
var gallery_1 = require("./gallery");
exports.Gallery = gallery_1.Gallery;
var gantt_1 = require("./gantt");
exports.Gantt = gantt_1.Gantt;
var html_editor_1 = require("./html-editor");
exports.HtmlEditor = html_editor_1.HtmlEditor;
var linear_gauge_1 = require("./linear-gauge");
exports.LinearGauge = linear_gauge_1.LinearGauge;
var list_1 = require("./list");
exports.List = list_1.List;
var load_indicator_1 = require("./load-indicator");
exports.LoadIndicator = load_indicator_1.LoadIndicator;
var load_panel_1 = require("./load-panel");
exports.LoadPanel = load_panel_1.LoadPanel;
var lookup_1 = require("./lookup");
exports.Lookup = lookup_1.Lookup;
var map_1 = require("./map");
exports.Map = map_1.Map;
var menu_1 = require("./menu");
exports.Menu = menu_1.Menu;
var multi_view_1 = require("./multi-view");
exports.MultiView = multi_view_1.MultiView;
var nav_bar_1 = require("./nav-bar");
exports.NavBar = nav_bar_1.NavBar;
var number_box_1 = require("./number-box");
exports.NumberBox = number_box_1.NumberBox;
var pie_chart_1 = require("./pie-chart");
exports.PieChart = pie_chart_1.PieChart;
var pivot_grid_1 = require("./pivot-grid");
exports.PivotGrid = pivot_grid_1.PivotGrid;
var pivot_grid_field_chooser_1 = require("./pivot-grid-field-chooser");
exports.PivotGridFieldChooser = pivot_grid_field_chooser_1.PivotGridFieldChooser;
var polar_chart_1 = require("./polar-chart");
exports.PolarChart = polar_chart_1.PolarChart;
var popover_1 = require("./popover");
exports.Popover = popover_1.Popover;
var popup_1 = require("./popup");
exports.Popup = popup_1.Popup;
var progress_bar_1 = require("./progress-bar");
exports.ProgressBar = progress_bar_1.ProgressBar;
var radio_group_1 = require("./radio-group");
exports.RadioGroup = radio_group_1.RadioGroup;
var range_selector_1 = require("./range-selector");
exports.RangeSelector = range_selector_1.RangeSelector;
var range_slider_1 = require("./range-slider");
exports.RangeSlider = range_slider_1.RangeSlider;
var recurrence_editor_1 = require("./recurrence-editor");
exports.RecurrenceEditor = recurrence_editor_1.RecurrenceEditor;
var resizable_1 = require("./resizable");
exports.Resizable = resizable_1.Resizable;
var responsive_box_1 = require("./responsive-box");
exports.ResponsiveBox = responsive_box_1.ResponsiveBox;
var sankey_1 = require("./sankey");
exports.Sankey = sankey_1.Sankey;
var scheduler_1 = require("./scheduler");
exports.Scheduler = scheduler_1.Scheduler;
var scroll_view_1 = require("./scroll-view");
exports.ScrollView = scroll_view_1.ScrollView;
var select_box_1 = require("./select-box");
exports.SelectBox = select_box_1.SelectBox;
var slide_out_1 = require("./slide-out");
exports.SlideOut = slide_out_1.SlideOut;
var slide_out_view_1 = require("./slide-out-view");
exports.SlideOutView = slide_out_view_1.SlideOutView;
var slider_1 = require("./slider");
exports.Slider = slider_1.Slider;
var sortable_1 = require("./sortable");
exports.Sortable = sortable_1.Sortable;
var sparkline_1 = require("./sparkline");
exports.Sparkline = sparkline_1.Sparkline;
var speed_dial_action_1 = require("./speed-dial-action");
exports.SpeedDialAction = speed_dial_action_1.SpeedDialAction;
var switch_1 = require("./switch");
exports.Switch = switch_1.Switch;
var tab_panel_1 = require("./tab-panel");
exports.TabPanel = tab_panel_1.TabPanel;
var tabs_1 = require("./tabs");
exports.Tabs = tabs_1.Tabs;
var tag_box_1 = require("./tag-box");
exports.TagBox = tag_box_1.TagBox;
var text_area_1 = require("./text-area");
exports.TextArea = text_area_1.TextArea;
var text_box_1 = require("./text-box");
exports.TextBox = text_box_1.TextBox;
var tile_view_1 = require("./tile-view");
exports.TileView = tile_view_1.TileView;
var toast_1 = require("./toast");
exports.Toast = toast_1.Toast;
var toolbar_1 = require("./toolbar");
exports.Toolbar = toolbar_1.Toolbar;
var tooltip_1 = require("./tooltip");
exports.Tooltip = tooltip_1.Tooltip;
var tree_list_1 = require("./tree-list");
exports.TreeList = tree_list_1.TreeList;
var tree_map_1 = require("./tree-map");
exports.TreeMap = tree_map_1.TreeMap;
var tree_view_1 = require("./tree-view");
exports.TreeView = tree_view_1.TreeView;
var validation_group_1 = require("./validation-group");
exports.ValidationGroup = validation_group_1.ValidationGroup;
var validation_summary_1 = require("./validation-summary");
exports.ValidationSummary = validation_summary_1.ValidationSummary;
var validator_1 = require("./validator");
exports.Validator = validator_1.Validator;
var vector_map_1 = require("./vector-map");
exports.VectorMap = vector_map_1.VectorMap;
